### ISSUE TITLE
Gemm fails if optional input C is not provided

### ISSUE THREAD

**Issue by georgen117** [microsoft/onnxruntime#8737]
**Describe the bug**
If you run Gemm without providing the optional C input it will fail to run saying `This is an invalid model`

Add this test to the gemm_test.cc unit test code
```
template <typename T>
void TestGemmNoTransNoCInput() {
  auto run_test = [](bool b_is_initializer) {
    OpTester test("Gemm");

    test.AddAttribute("transA", (int64_t)0);
    test.AddAttribute("transB", (int64_t)0);
    test.AddAttribute("alpha", 1.0f);
    test.AddAttribute("beta", 1.0f);

    test.AddInput<T>("A", {2, 4},
                     {1.0f, 2.0f, 3.0f, 4.0f,
                      -1.0f, -2.0f, -3.0f, -4.0f});
    test.AddInput<T>("B", {4, 3}, std::vector<T>(12, 1.0f), b_is_initializer);
    test.AddOutput<T>("Y", {2, 3},
                      {10.0f, 10.0f, 10.0f,
                       -10.0f, -10.0f, -10.0f});
    test.Run();
  };

  run_test(false);
  // NNAPI EP requires weight to be an initializer
  run_test(true);
}

TEST(GemmOpTest, GemmNoTransNoCInput_float) {
  TestGemmNoTransNoCInput<float>();
}

TEST(GemmOpTest, GemmNoTransNoCInput_double) {
  TestGemmNoTransNoCInput<double>();
}
```

This is a modification of an already existing test I simply removed the C input.

According to the ONNX specification the 3rd input is optional
https://github.com/onnx/onnx/blob/master/docs/Operators.md#Gemm

So the Gemm operator should still run.

**Urgency**
To the best of my knowledge this is not blocking anyone.
The issue seen here may be related to another issue reported by me: https://github.com/microsoft/onnxruntime/issues/7506

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- ONNX Runtime installed from (source or binary): built from source
- ONNX Runtime version: 1.8.1 (close to the tip of master)
- Python version: Python 3.8.3
- Visual Studio version (if applicable):  Visual Sudio Pro 2019 Version 4.8.04084
- GCC/Compiler version (if compiling from source): N/A (not yet tried issue on linux)
- CUDA/cuDNN version: N/A using CPU build
- GPU model and memory: from unit test see code above

**To Reproduce**
- Describe steps/code to reproduce the behavior.

Copy pase unit test code given above into `onnxruntime\test\providers\cpu\math\gemm_test.cc`
build the code: `.\build.bat --config RelWithDebInfo --parallel  --build_shared_lib --cmake_generator "Visual Studio 16 2019"`
run the unit tests in quesiton
`.\build\Windows\RelWithDebInfo\RelWithDebInfo\onnxruntime_test_all.exe --test_filter=GemmOpTest.GemmNoTransNoCInput*`

**Expected behavior**
The two tests should run and pass without failure.

**Actual behavior**
```
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from GemmOpTest
[ RUN      ] GemmOpTest.GemmNoTransNoCInput_float
2021-08-13 13:43:23.6822803 [E:onnxruntime:Default, provider_test_utils.cc:960 onnxruntime::test::OpTester::Run] Resolve failed with status: This is an invalid model. Error in Node:node1 : Node (node1) has input size 2 not in range [min=3, max=3].
C:\onnxruntime\onnxruntime\test\providers\provider_test_utils.cc(962): error: Value of: status.IsOK()
  Actual: false
Expected: true
This is an invalid model. Error in Node:node1 : Node (node1) has input size 2 not in range [min=3, max=3].
2021-08-13 13:43:23.6870943 [E:onnxruntime:Default, provider_test_utils.cc:960 onnxruntime::test::OpTester::Run] Resolve failed with status: This is an invalid model. Error in Node:node1 : Node (node1) has input size 2 not in range [min=3, max=3].
C:\onnxruntime\onnxruntime\test\providers\provider_test_utils.cc(962): error: Value of: status.IsOK()
  Actual: false
Expected: true
This is an invalid model. Error in Node:node1 : Node (node1) has input size 2 not in range [min=3, max=3].
[  FAILED  ] GemmOpTest.GemmNoTransNoCInput_float (11 ms)
[ RUN      ] GemmOpTest.GemmNoTransNoCInput_double
2021-08-13 13:43:23.6919917 [E:onnxruntime:Default, provider_test_utils.cc:960 onnxruntime::test::OpTester::Run] Resolve failed with status: This is an invalid model. Error in Node:node1 : Node (node1) has input size 2 not in range [min=3, max=3].
C:\onnxruntime\onnxruntime\test\providers\provider_test_utils.cc(962): error: Value of: status.IsOK()
  Actual: false
Expected: true
This is an invalid model. Error in Node:node1 : Node (node1) has input size 2 not in range [min=3, max=3].
2021-08-13 13:43:23.6950867 [E:onnxruntime:Default, provider_test_utils.cc:960 onnxruntime::test::OpTester::Run] Resolve failed with status: This is an invalid model. Error in Node:node1 : Node (node1) has input size 2 not in range [min=3, max=3].
C:\Users\georgena\workspace\onnxruntime\onnxruntime\test\providers\provider_test_utils.cc(962): error: Value of: status.IsOK()
  Actual: false
Expected: true
This is an invalid model. Error in Node:node1 : Node (node1) has input size 2 not in range [min=3, max=3].
[  FAILED  ] GemmOpTest.GemmNoTransNoCInput_double (13 ms)
[----------] 2 tests from GemmOpTest (26 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (33 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] GemmOpTest.GemmNoTransNoCInput_float
[  FAILED  ] GemmOpTest.GemmNoTransNoCInput_double

 2 FAILED TESTS
```

**Comment by kit1980** [comment]
@georgen117
Looks like the same cause as in https://github.com/microsoft/onnxruntime/issues/7506: opset version is before Gemm's 3rd input became optional.

OpTester defaults to opset 7 (https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/test/providers/provider_test_utils.h#L258), and for opset 7 the model is indeed invalid.

The test should specify the opset: `OpTester test("Gemm", 10);`

**Comment by georgen117** [comment]
@kit1980 
> The test should specify the opset: OpTester test("Gemm", 10);

I updated the code to `OpTester test("Gemm", 10);` and it still failed.

I then checked the Onnx specification the "C" input was not changed to optional till op-set 11. So I updated the test to 

`OpTester test("Gemm", 11);`  The test passed!  I am going to close this issue as it is not a problem when the op-set version is correctly set.  It is possible that the optional input is still affecting the GemmGrad op for older op-set versions but that issue has been reported in #7506.
